user/pcmcia-cs/man/setpnp.8
.\" Copyright (C) 1999 David A. Hinds -- dahinds@users.sourceforge.net
.\" setpnp.8 1.3 2000/06/12 21:24:49
.\"
.TH SETPNP 8 "2000/06/12 21:24:49" "pcmcia-cs"
.SH NAME
setpnp \- modify Plug and Play BIOS device resources
.SH SYNOPSIS
.B setpnp
.RB [ -b ]
.I device resource...
.br
.B setpnp
.RB [ -b ]
.I device {on|off}
.SH DESCRIPTION
This utility updates the system resource information for Plug and Play
BIOS device nodes.  A device is specified by its two-digit hex node
number.  In its first form, a list of resources are given on the
command line.  Alternatively, a device can either be simply switched
``on'' (reset to its boot configuration) or ``off'' (all resources
disabled).
.PP
There are four types of resources: ``io'', ``mem'', ``irq'', and
``dma''.  A resource list consists of one or more space-separated
pairs of a resource type with one or more values.  Multiple values for
one resource type can be separated by commas.  Values can also either
be unitary or dash-separated ranges.  For example:
.sp
.RS
.nf
setpnp 0d io 0x2f8-0x2ff irq 3
setpnp 12 io 0x350-0x35f,0x2f8-0x2ff irq 3 irq 10
.RE
.fi
.sp
The order of items of different types is not important, but if more
than one item of the same type is present, their relative order is
significant.  The Plug and Play BIOS will reject invalid configuration
attempts; however,
.B setpnp
will not attempt to determine why a configuration was rejected.
.PP
By default, current (dynamic) device configuration information is
modified.  With the
.B -b
option, a device's boot (static) configuration can be updated.  Some
devices may only be reconfigured for the following boot.  Be especially
careful when modifying your system's boot configuration.  Improper use
of this command may disable vital system devices and render your
system unbootable. 
.SH OPTIONS
.TP
.B \-b
Boot mode: update the device resource information that will be used at
next boot (as opposed to current resource info).
.TP
.SH FILES
.TP \w'/proc/bus/pnp/...\ \ \ \ |\|'u
/proc/bus/pnp/...
The kernel interface for Plug and Play BIOS device services.
.SH AUTHORS
David Hinds \- dahinds@users.sourceforge.net
.SH "SEE ALSO"
lspnp(8)


freeswan/CHANGES
New and notable in 1.97:

Sanity check added for at least one of AH or ESP more for checking if
"make *config" has even been run.

Fixed 2.2 local IKE fragmentation blackhole.  Still won't work if
iptraf or another pcap app is running.

MCR added ikeping(8).

Applied DHR's tunnel patch to streamline IKE/specialSA processing.

Added local and remote IKE bypass tests to the UML test suite.

UML KLIPS %trap* tests now check for upbound pf_key messages.

An ICMP ICMP_DEST_UNREACH, ICMP_PKT_FILTERED is now sent upon %reject.

Added a UML input file with first and last packets being outside the 
covered range to "sandwich" the packets under test.

The default _updown script has been cleaned up slightly, reducing
internal duplication that arose accidentally.

When doing --add on an RSA-signature conn, ipsec_auto now tells Pluto
about the keys before, not after, telling it about the conn.  The main
consequence of this is that if something is wrong with one of the keys,
the conn itself does *not* get loaded.  This should reduce confusion.

The rp_filter checks during KLIPS startup have been fixed to be a bit less
paranoid; rp_filter is thought to cause problems only on the physical
interfaces, not on the ipsecN interfaces. 

The configuration process has been fixed to be a bit more paranoid.  Some
users have had mysterious compile failures with complaints involving names
like MD5_CTX; they seem to have resulted from kernel configuration changes
somehow getting into .config but not into include/linux/autoconf.h (which
is what actually controls most of the kernel compile).  The post-config
checking now verifies not only that .config has the right stuff in it, but
that autoconf.h does too.  (We still don't see *how* this discrepancy can
arise, but this will at least catch it when it does.)

The details of linking the KLIPS source into the kernel are now under
control of a configuration variable in Makefile.inc, to make it easier
for distribution builders to change.

The top-level Makefile now includes cleverness which assigns a date-based
version code if a build is done from a raw CVS checkout.

The library now includes a new function, ttodatav(), like ttodata() but
in some cases it gives more verbose error reports.  Also, there are now
a couple of functions which distill an RSA public key to a "key ID" for
use in messages, a crypto-quality pseudo-random-number generator, and a
function that converts a single address into a (singleton) subnet.

Barf's secret/key-censoring utilities have been tidied up so that what
they emit matches the key-ID library functions. 

The ipsec command now supplies a couple more environment variables to
things it calls, and this has reduced the number of things which need to
be edited at build time to know about configuration variables. 

The configuration reader now allows parameter names to begin with "_";
these are strictly reserved for internal development purposes and will
not in general be documented.


As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.96:

Compatibility with the 2.0.xx kernels (under Red Hat 5.2) is broken.
This was accidental, and we have some hopes of fixing it, although it
won't last forever even if we do.  See README.

Our "IKE Implementation Issues" IETF draft has been added to our docs.

Newhostkey now has a mandatory --output "option" specifying a filename
(although the magic filename "-" means "standard output"), and sets
restrictive permissions on newly-created files to try to keep private keys
truly private. 

Pluto now gracefully handles a couple of odd error cases (missing private
keys, network errors at just the wrong time in negotiations) which used to
cause it to die, provoking an automatic restart of the IPsec subsystem. 

Makefile.inc now has more detailed control over where manpages go, to ease
integration into distributions.  The name of the destination directory for
RPMs is also now centralized there.

Rsasigkey now checks for certain obscure internal failures, which can
happen if --random is used to supply a source of random bits that aren't
really very random, and complains accordingly.

Barf's eroute table, formerly sorted by source, is now sorted by destination.
This is probably a bit clearer for hosts with lots of eroutes.

Use of the %reject target in a shunt eroute now, as long planned, actually
causes an ICMP rejection message (Destination Unreachable:  Communication
Administratively Prohibited). 

KLIPS now declares its module license to be "Dual BSD/GPL".  (It is 
dual because the libdes and zlib code is BSD licensed, while KLIPS itself 
is GPL).  Whether this is exactly right is not yet clear, but at least it
stops the latest module loaders from whining.

The UML test stuff continues to grow and add new test cases.

As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.95:

The last remnants of the "%hold" bug, which broke 1.93 and 1.94, have
(we think) been dealt with.

Minor bugs in the RPM-building stuff have been fixed, and it has been
updated to deal with Red Hat's abrupt removal of the --buildroot option
from rpm.

Pluto now supports the ability to delete an instance of a conn, rather
than just the whole conn (for cases, notably OE, where a single connection
description can be involved in multiple connections).  There is no support
for this in ipsec_auto yet.

Pluto's routine (non-debug) logging is now somewhat terser, and some of
the more boring messages are produced only if --debug-lifecycle is turned
on (this is a temporary hack, it's not really debugging output).

Pluto's innards have been reorganized somewhat, in preparation for making
DNS lookups asynchronous.  There should be no user-visible effects.

Pluto now ignores IPv6 link-local interface addresses; we think that they
are never relevant, and trying to bind to them causes difficulties.  This

user/pcmcia-cs/debug-tools/setpnp.c
/*======================================================================

    A utility for reconfiguring PnP BIOS devices

    setpnp.c 1.7 2000/06/12 21:34:19

    The contents of this file are subject to the Mozilla Public
    License Version 1.1 (the "License"); you may not use this file
    except in compliance with the License. You may obtain a copy of
    the License at http://www.mozilla.org/MPL/

    Software distributed under the License is distributed on an "AS
    IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
    implied. See the License for the specific language governing
    rights and limitations under the License.

    The initial developer of the original code is David A. Hinds
    <dahinds@users.sourceforge.net>.  Portions created by David A. Hinds
    are Copyright (C) 1999 David A. Hinds.  All Rights Reserved.

    Alternatively, the contents of this file may be used under the
    terms of the GNU Public License version 2 (the "GPL"), in which
    case the provisions of the GPL are applicable instead of the
    above.  If you wish to allow the use of your version of this file
    only under the terms of the GPL and not to allow others to use
    your version of this file under the MPL, indicate your decision
    by deleting the provisions above and replace them with the notice
    and other provisions required by the GPL.  If you do not delete
    the provisions above, a recipient may use your version of this
    file under either the MPL or the GPL.

    Usage:

    setpnp [-b] [device #] [resource list]
    setpnp [-b] [device #] {on|off}

    The device number is a two-digit hex string.  The resource list
    consists of a series of resource names and values.  Four resource
    names are available: "io", "mem", "irq", and "dma".  Values can
    either be single numbers or dash-delimited ranges.  More than one
    value can be listed in a single argument, separated by commas.
    
    For example:

    setpnp 0d irq 3 io 0x02f8-0x02ff

======================================================================*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>
#include <endian.h>
#include <ctype.h>
#include <asm/types.h>

#include <linux/pnp_resource.h>

static int verbose = 0, boot = 0;

#define swap16(n) ((((n)&0x00ff)<<8) | (((n)&0xff00)>>8))
#define swap32(n) \
    ((((n)&0xff000000)>>24) | (((n)&0x00ff0000)>>8) | \
     (((n)&0x0000ff00)<<8)  | (((n)&0x000000ff)<<24))

#if (__BYTE_ORDER == _BIG_ENDIAN)
#define flip16(n)	swap16(n)
#define flip32(n)	swap32(n)
#else
#define flip16(n)	(n)
#define flip32(n)	(n)
#endif

#define NRSRC	4
#define NBASE	8
#define R_IO	0
#define R_MEM	1
#define R_IRQ	2
#define R_DMA	3

struct rsrc_list {
    int		nr[NRSRC];
    u_long	base[NRSRC][NBASE];
    u_long	len[NRSRC][NBASE];
};

static const char *rsrc_type[] = { "io", "mem", "irq", "dma" };

/*====================================================================*/

static char *update_chain(u_char *buf, int nr, struct rsrc_list *res)
{
    union pnp_resource *p = (union pnp_resource *)buf;
    int tag = 0, sz, nu[4];
    u_long base, len;
    
    nu[0] = nu[1] = nu[2] = nu[3] = 0;
    while (((u_char *)p < buf+nr) && (tag != PNP_RES_SMTAG_END)) {
	if (p->lg.tag & PNP_RES_LARGE_ITEM) {
	    union pnp_large_resource *r = &p->lg.d;
	    tag = p->lg.tag & ~PNP_RES_LARGE_ITEM;
	    sz = flip16(p->lg.sz) + 2;
	    switch (tag) {
	    case PNP_RES_LGTAG_MEM:
		if (res->nr[R_MEM] > nu[R_MEM]) {
		    base = res->base[R_MEM][nu[R_MEM]++];
		    len = res->len[R_MEM][nu[R_MEM]++];
		    r->mem.min = r->mem.max = flip16(base >> 8);
		    r->mem.len = flip16(len);
		}
		break;
	    case PNP_RES_LGTAG_MEM32:
		if (res->nr[R_MEM] > nu[R_MEM]) {
		    base = res->base[R_MEM][nu[R_MEM]++];
		    len = res->len[R_MEM][nu[R_MEM]++];
		    r->mem32.min = r->mem32.max = flip32(base);
		    r->mem32.len = flip32(len);
		}
		break;
	    case PNP_RES_LGTAG_MEM32_FIXED:
		if (res->nr[R_MEM] > nu[R_MEM]) {
		    base = res->base[R_MEM][nu[R_MEM]];
		    len = res->len[R_MEM][nu[R_MEM]++];
		    r->mem32_fixed.base = flip32(base);
		    r->mem32_fixed.len = flip32(len);
		}
		break;
	    }
	} else {
	    union pnp_small_resource *r = &p->sm.d;
	    tag = (p->sm.tag >> 3); sz = (p->sm.tag & 7);
	    switch (tag) {
	    case PNP_RES_SMTAG_IRQ:
		if (res->nr[R_IRQ] > nu[R_IRQ]) {
		    base = res->base[R_IRQ][nu[R_IRQ]++];
		    r->irq.mask = base ? flip16(1<<base) : 0;
		}
		break;
	    case PNP_RES_SMTAG_DMA:
		if (res->nr[R_DMA] > nu[R_DMA]) {
		    base = res->base[R_DMA][nu[R_DMA]++];
		    r->dma.mask = base ? flip16(1<<base) : 0;
		}
		break;
	    case PNP_RES_SMTAG_IO:
		if (res->nr[R_IO] > nu[R_IO]) {
		    base = res->base[R_IO][nu[R_IO]];
		    len = res->len[R_IO][nu[R_IO]++];
		    r->io.min = r->io.max = flip16(base);
		    r->io.len = len;
		}
		break;
	    case PNP_RES_SMTAG_IO_FIXED:
		if (res->nr[R_IO] > nu[R_IO]) {
		    base = res->base[R_IO][nu[R_IO]++];
		    len = res->len[R_IO][nu[R_IO]++];
		    r->io_fixed.base = flip16(base);
		    r->io_fixed.len = len;
		}
		break;
	    }
	}
	(u_char *)p += sz + 1;
    }
    return (u_char *)p;
}

static int update_resources(int num, struct rsrc_list *res)
{
    char fn[40];
    u_char buf[4096];
    int fd, nr, nw;
    
    if (access("/proc/bus/pnp", F_OK) != 0) {
	fprintf(stderr, "lspnp: /proc/bus/pnp not available\n");
	return EXIT_FAILURE;
    }
    
    sprintf(fn, "/proc/bus/pnp/%s%02x", (boot ? "boot/" : ""), num);
    fd = open(fn, O_RDWR);
    nr = read(fd, buf, sizeof(buf));
    if (nr <= 0) {
	perror("read failed");
	return EXIT_FAILURE;
    }
    
    update_chain(buf, nr, res);
    nw = write(fd, buf, nr);
    close(fd);
    if (nr != nw) {
	perror("write failed");
	return EXIT_FAILURE;
    }
    return EXIT_SUCCESS;
}

static int reset_resources(int num)
{
    char fn[40];
    u_char buf[4096];
    int fd, nr, nw;
    
    if (access("/proc/bus/pnp", F_OK) != 0) {
	fprintf(stderr, "lspnp: /proc/bus/pnp not available\n");
	return EXIT_FAILURE;
    }
    sprintf(fn, "/proc/bus/pnp/boot/%02x", num);
    fd = open(fn, O_RDONLY);
    nr = read(fd, buf, sizeof(buf));
    close(fd);
    if (nr <= 0) {
	perror("read failed");
	return EXIT_FAILURE;
    }
    sprintf(fn, "/proc/bus/pnp/%02x", num);
    fd = open(fn, O_WRONLY);
    nw = write(fd, buf, nr);
    close(fd);
    if (nr != nw) {
	perror("write failed");
	return EXIT_FAILURE;
    }
    return EXIT_SUCCESS;
}

/*====================================================================*/

static int parse_resources(char *argv[], int argc,
			   struct rsrc_list *res)
{
    int i, j;
    u_long base, len;
    char *s, *t;
    
    for (i = 0; i < argc; i += 2) {
	for (j = 0; j < NRSRC; j++)
	    if (strcmp(rsrc_type[j], argv[i]) == 0) break;
	if (j == NRSRC) {
	    fprintf(stderr, "bad resource type: '%s'\n", argv[i]);
	    return EXIT_FAILURE;
	}
	s = strtok(argv[i+1], ", \t");
	while (s) {
	    base = strtoul(s, &t, 0);
	    len = ((*t == '-') ? strtoul(t+1, &t, 0)-base+1 : 1);
	    if ((*s == '\0') || (*t != '\0')) {
		fprintf(stderr, "bad resource argument: '%s'\n", t);
		return EXIT_FAILURE;
	    }
	    res->base[j][res->nr[j]] = base;
	    res->len[j][res->nr[j]++] = len;
	    s = strtok(NULL, ", \t");
	}
    }
    return EXIT_SUCCESS;
}

/*====================================================================*/

void usage(char *name)
{
    fprintf(stderr, "usage: %s [-b] [device #] [resources ...]\n"
	    "    or %s [-b] [device #] {on|off}\n", name, name);
    exit(EXIT_FAILURE);
}
    
int main(int argc, char *argv[])
{
    int i, optch, errflg = 0;
    static struct rsrc_list res;
    char *s;
    
    while ((optch = getopt(argc, argv, "bv")) != -1) {
	switch (optch) {
	case 'b':
	    boot++; break;
	case 'v':
	    verbose++; break;
	default:
	    errflg = 1; break;
	}
    }
    if (errflg || (optind == argc))
	usage(argv[0]);
    
    i = strtoul(argv[optind], &s, 16);
    if ((argv[optind] == '\0') || (*s != '\0'))
	usage(argv[0]);
    optind++;

    /* Special commands */
    if (argc == optind+1) {
	if (strcmp(argv[optind], "off") == 0) {
	    res.nr[0] = res.nr[1] = res.nr[2] = res.nr[3] = 7;
	    optind++;
	} else if (strcmp(argv[optind], "on") == 0) {
	    return reset_resources(i);
	} else {
	    usage(argv[0]);
	}
    } else if (argc == optind)
	usage(argv[0]);
    
    if ((argc - optind) % 2)
	usage(argv[0]);
    if (parse_resources(argv+optind, argc-optind, &res) == 0)
	return update_resources(i, &res);
    return EXIT_FAILURE;
}
